subsys/bluetooth/host/buf.c

```diff
@@ -147,3 +151,62 @@ struct net_buf_pool *bt_buf_get_num_complete_pool(void)
 }
 #endif /* CONFIG_BT_CONN || CONFIG_BT_ISO */
 #endif /* ZTEST_UNITTEST */
+
+struct net_buf *bt_buf_make_view(struct net_buf *view,
+				 struct net_buf *parent,
+				 size_t len,
+				 struct bt_buf_view_meta *meta)
+{
+	__ASSERT_NO_MSG(len);
+	__ASSERT_NO_MSG(view);
+	/* The whole point of this API is to allow prepending data. If the
+	 * headroom is 0, that will not happen.
+	 */
+	__ASSERT_NO_MSG(net_buf_headroom(parent) > 0);
+
+	__ASSERT_NO_MSG(!bt_buf_has_view(parent));
+
+	LOG_DBG("make-view %p viewsize %u meta %p", view, len, meta);
+
+	net_buf_simple_clone(&parent->b, &view->b);
+	view->size = net_buf_headroom(parent) + len;
+	view->len = len;
+	view->flags = NET_BUF_EXTERNAL_DATA;
+
+	/* we have a view, eat `len`'s worth of data from the parent */
+	(void)net_buf_pull(parent, len);
+
+	meta->backup.data = parent->data;
+	parent->data = NULL;
+
+	meta->backup.size = parent->size;
+	parent->size = 0;
+
+	/* The ref to `parent` is moved in by passing `parent` as argument. */
+	/* save backup & "clip" the buffer so the next `make_view` will fail */
+	meta->parent = parent;
```

Comment:
Maybe I just forgot what we discussed & agreed about this, but wouldn't it be cleaner to use a pointer-to-pointer for any new APIs which move the reference, and set the original pointer to NULL as part of this operation?

I get the impression that `bt_buf_destroy_view()` is supposed to "restore" the parent, in which case wouldn't it make sense to return back the original parent buffer as a return value of `bt_buf_destroy_view()`?

Reply:
We modeled it after `net_buf_destroy()` which returns a void.

Not returning the parent encourages not calling synchronous code from the `destroy()` callback of the view-buffer pool, and that's the only place `bt_buf_destroy_view()` should get called.

Anyway, `buf_view.h` is private so we can change the API as we wish in a later PR, taking more time to discuss better designs. 

Another thing I'd like to discuss then is how to do some optimizations, like passing the original `buf` directly if it fits the HCI size in a easy-to-understand way.

Reply:
@jori-nordic makes sense, thanks for the reminder & clarification